packages/playwright-core/src/server/dispatchers/browserDispatcher.ts

```diff
@@ -47,24 +59,49 @@ export class BrowserDispatcher extends Dispatcher<Browser, channels.BrowserChann
   }
 
   async newContext(params: channels.BrowserNewContextParams, metadata: CallMetadata): Promise<channels.BrowserNewContextResult> {
+    if (!this._options.isolateContexts) {
+      const context = await this._object.newContext(metadata, params);
+      const contextDispatcher = BrowserContextDispatcher.from(this, context);
+      return { context: contextDispatcher };
+    }
+
+    if (params.recordVideo)
+      params.recordVideo.dir = this._object.options.artifactsDir;
     const context = await this._object.newContext(metadata, params);
-    return { context: BrowserContextDispatcher.from(this, context) };
+    this._isolatedContexts.add(context);
+    context.on(BrowserContext.Events.Close, () => this._isolatedContexts.delete(context));
+    const contextDispatcher = BrowserContextDispatcher.from(this, context);
+    this._dispatchEvent('context', { context: contextDispatcher });
+    return { context: contextDispatcher };
   }
 
   async newContextForReuse(params: channels.BrowserNewContextForReuseParams, metadata: CallMetadata): Promise<channels.BrowserNewContextForReuseResult> {
-    return await newContextForReuse(this._object, this, params, metadata);
+    const { context, needsReset } = await this._object.newContextForReuse(params, metadata);
+    if (needsReset) {
+      const oldContextDispatcher = this.connection.existingDispatcher<BrowserContextDispatcher>(context);
+      if (oldContextDispatcher)
+        oldContextDispatcher._dispose();
+      await context.resetForReuse(metadata, params);
+    }
+    const contextDispatcher = BrowserContextDispatcher.from(this, context);
+    this._dispatchEvent('context', { context: contextDispatcher });
+    return { context: contextDispatcher };
   }
 
   async stopPendingOperations(params: channels.BrowserStopPendingOperationsParams, metadata: CallMetadata): Promise<channels.BrowserStopPendingOperationsResult> {
     await this._object.stopPendingOperations(params.reason);
   }
 
   async close(params: channels.BrowserCloseParams, metadata: CallMetadata): Promise<void> {
+    if (this._options.ignoreStopAndKill)
```

Comment:
Current behavior "In case this browser is connected to, clears all created contexts belonging to this browser and disconnects from the browser server."

Maybe we should also close all contexts created by this connection?

Reply:
We do that in `cleanupContexts()` method.